Relations of fundamental invariants

joschmitt opened the pull request:
I implemented an algorithm by Kemper and Steel to compute the relations between fundamental invariants when they are the union of primary and irreducible secondary invariants which relies on linear algebra and not Gröbner bases.

joschmitt committed:
First go at relations / algebra syzygies for invariant rings

joschmitt committed:
Speed up linear algebra with polynomials

joschmitt committed:
Rework relations of invariants

joschmitt committed:
`_containement_helper` -> `_containment_helper`

joschmitt committed:
Add homogeneous version of `subalgebra_membership`

joschmitt committed:
Rework `fundamental_invariants_via_primary_and_secondary`:
Save the representation of removed invariants in the generators

joschmitt committed:
Tie the new `relations` function to `affine_algebra`

fingolfin commented on src/Deprecations.jl:
```diff
@@ -1,3 +1,4 @@
 # Deprecated in 0.10.*
 
 @deprecate automorphisms(x::Graph) automorphism_group_generators(x)
+@deprecate _containement_helper(R::MPolyRing, N::Int, M::Int, I::MPolyIdeal, W::Vector, ord::Symbol) _containment_helper(R, N, M, I, W, ord)
```
This is an internal only function, isn't it? So I don't think we need to deprecate it; or do you have reason to believe there's external code using it?

joschmitt replied:
I just thought it might be best practice. But probably it is overly careful. I remove the `@deprecate`.

fingolfin commented on src/InvariantTheory/affine_algebra.jl:
```diff
@@ -0,0 +1,249 @@
+invariants is used. With `:groebner_basis`, the relations are computed via the
+standard computation of a kernel of a morphism between multivariate polynomial
+rings. The option `:linear_algebra` uses an algorithm by Kemper and Steel
+[KS99, Section 17.5.5] to compute the relations without the use of Groebner
```
Could you perhaps change this to use `@cite` (grep through the `docs` dir to see examples; the bib file already contains KS99)

joschmitt replied:
Yes, thanks, I totally forgot that.

fingolfin reviewed (approved):
Thank you, looks great!

(Two minor suggestions, not important, feel free to ignore)

fingolfin commented:
Also, the documentation test fail. You could try `Oscar.build_doc(doctest = :fix)` (must be done with Julia 1.6)

joschmitt committed:
Fix doc tests and minor changes

thofma commented:
Do you mind adding a test where we can quickly check the result? Like a symmetric group:
```
M1 = matrix(QQ, 4, 4, [0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0])
M2 = matrix(QQ, 4, 4, [0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1])
R = invariant_ring(M1, M2)
A, = affine_algebra(R)
@test is_zero(modulus(A))
```

Is there a specific reason the `affine_algebra` function moved? Now the tests are in the "wrong" place.

joschmitt committed:
Fix corner case and more tests

joschmitt commented:
I added more tests (including an example from the paper by Kemper and Steel) and fixed a corner case...

> Is there a specific reason the `affine_algebra` function moved? Now the tests are in the "wrong" place.

I personally prefer several small files to one large file and with the code for the relations it felt justified to open a new file. I don't see which tests are in the wrong place now. `affine_algebra` was not tested before, I think?

thofma commented:
Sorry, I misread something. Everything is fine with respect to the files. My bad.

thofma commented:
Thanks @joschmitt!
